nixos/modules/config/sysctl.nix

```diff
@@ -59,5 +59,19 @@ in
     # Disable YAMA by default to allow easy debugging.
     boot.kernel.sysctl."kernel.yama.ptrace_scope" = mkDefault 0;
 
+    # Dynamic since kernel v5.12
+    # https://github.com/torvalds/linux/commit/ac7b79fd190b02e7151bc7d2b9da692f537657f3
+    boot.kernel.sysctl."fs.inotify.max_user_instances" =
+      if (config.boot.kernelPackages.kernel.kernelAtLeast "5.12")
+      then mkDefault 2147483647 # INT_MAX, dynamically limited based on available memory
+      else mkDefault 8192; # instead of 128
+
+    # Dynamic since kernel v5.11
+    # https://github.com/torvalds/linux/commit/92890123749bafc317bbfacbe0a62ce08d78efb7
+    boot.kernel.sysctl."fs.inotify.max_user_watches" =
+      mkIf
+        (config.boot.kernelPackages.kernel.kernelAtLeast "5.11")
+        (mkDefault 1048576); # instead of 8192
```

Comment:
Oh, absolutely right: https://github.com/NixOS/nixpkgs/compare/de892e559efb0a3e93ddb8799f6d113f178bf3b4..f47bf35286bb8f968f1a717b12c7f25f603ba10b

Reply:
That didn't fix it. How can I express that with nix appropriately?

Reply:
```nix
{
    boot.kernel.sysctl."fs.inotify.max_user_watches" = mkIf
      !(config.boot.kernelPackages.kernel.kernelAtLeast "5.11")
      (mkDefault 1048576); # instead of 8192
}
```

Is that good? That is: doesn't set `fs.inotify.max_user_watches` at all for kernel versions that are not at least 5.11?

Reply:
You should use `kernelOlder` here instead of inverting `kernelAtLeast`, but the logic here looks right to me.

If you want to be extra sure you can test with `nixos-rebuild build-vm`.

Reply:
I suppose you mean a hypothetical `kernelYounger "5.11"`, but I checked that doesn't exists :disappointed:

Reply:
```nix
versionAtLeast = v1: v2: !versionOlder v1 v2
```

`kernelOlder` should do exactly what you want.

Reply:
The default should apply for any `kernelVersion < 5.11`. That would be `kernelYounger`.

The suggestions about `versionOlder` / `kernelOlder`do confuse me. How are the to be parsed (in plain English)?

Reply:
The following does what you want, right?

```nix
!(config.boot.kernelPackages.kernel.kernelAtLeast "5.11")
```

This is equivalent to

```nix
!(!(config.boot.kernelPackages.kernel.kernelAtLeast "5.11"))
```

and we can eliminate the negations.

`kernelOlder v` is not `kernelVersion <= v`, but `kernelVersion < v` which is what you seem to want to call `kernelYounger`?

Reply:
I'm getting completely lost, sry for that.

> kernelOlder v is not kernelVersion <= v, but kernelVersion < v which is what you seem to want to call kernelYounger?

Doe you mean `kernelOlder` is _actually_ `kernelYounger` and just horribly named? `kernelOlder` would be `kernelVersion > v` not `kernelVersion >= v`. And herein lies the problem:

---

- `[5.11 - <infinity>` - no defaults
- `-<infinity> - 5.11[` - defaults

If `kernelAtLeast 5.12` _is_ `[5.11 - <infinity>`, then the negation of that is everything else. That everything else is where we want to apply the defaults.

Reply:
Seems to be doing what you want:

```
nix-repl> linuxPackages_5_12.kernelOlder "5.11"
false

nix-repl> linuxPackages_5_10.kernelOlder "5.11" 
true
```

`kernelOlder` means “Is this kernel's version older than the provided one?” and is equivalent to:

```nix
lib.versionOlder kernel.version versionToCompareTo
```

---

> If kernelAtLeast 5.11 is [5.11 - <infinity>, then the negation of that is everything else. That everything else is where we want to apply the defaults.

`kernelOlder` is precisely the negation of `kernelAtLeast` (see <https://github.com/NixOS/nixpkgs/pull/126777#discussion_r654837656>), so exactly what you want: The “older” means version lower than, so it doesn *not* include the provide version:

```
nix-repl> let kernelPackageVersion = "5.11"; in lib.versionOlder kernelPackageVersion "5.11"
false
```

Reply:
@sternenseemann Thanks for the "mental lift", it definitely helped. But this is still not correct:

```
# This is not what we want
nix-repl> linuxPackages_5_11.kernelOlder "5.11"  # there is no 5_11, but you can try with 5.10
false

# This is what we want
nix-repl> linuxPackages_5_11.kernelAtLeast "5.11"  # there is no 5_11, but you can try with 5.10
true
```

Reply:
I thought you wanted a predicate that would exclude 5.11? If you negate kernelAtLeast then you still get the same result as with kernelOlder.

Reply:
Finally it clicked, thanks @sternenseemann !

Resolved in : https://github.com/NixOS/nixpkgs/compare/8c967c1942645b739d472f8b1a495c1aae462719..0e8af3ed9687017ba06735c9220b280793b14d06